[RPC Framework] Replace Python Pickler with internal RPC pickler for RemoteModule

wayi1 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#58019 [RPC Framework] Replace Python Pickler with internal RPC pickler for RemoteModule**

In order to support sending `RemoteModule` over PRC, previously the pickling/unpickling of `RemoteModule` was implemented based on `__setstate__` and `__getstate__`. However, this means that the user can call regular Python pickler/unpickler to invoke the same logic,which should not be allowed.

This PR ensures that the pickling can only happen over RPC and not via regular python pickle.

Additionally, when a new attribute is added to `RemoteModule`, if it's not added to either `_REMOTE_MODULE_PICKLED_ATTRIBUTES` or `_REMOTE_MODULE_ATTRIBUTES_IGNORE_FOR_PICKLING`, this attribute will be ignored and an error message will be printed to std.err. However, it will not raise an exception like before, because such exception raised at the RPC layer will somehow cause timeout.

#Closes: https://github.com/pytorch/pytorch/issues/57516

Differential Revision: [D28318270](https://our.internmc.facebook.com/intern/diff/D28318270/)

Commit:
[RPC Framework] Replace Python Pickler with internal RPC pickler for RemoteModule

In order to support sending `RemoteModule` over PRC, previously the pickling/unpickling of `RemoteModule` was implemented based on `__setstate__` and `__getstate__`. However, this means that the user can call regular Python pickler/unpickler to invoke the same logic,which should not be allowed.

This PR ensures that the pickling can only happen over RPC and not via regular python pickle.

Additionally, when a new attribute is added to `RemoteModule`, if it's not added to either `_REMOTE_MODULE_PICKLED_ATTRIBUTES` or er`_REMOTE_MODULE_ATTRIBUTES_IGNORE_FOR_PICKLING`, this attribute will be ignored and an error message will be printed to std.err. However, it will not raise an exception like before, because such exception raised at the RPC layer won't be displayed properly.

#Closes: https://github.com/pytorch/pytorch/issues/57516

Differential Revision: [D28318270](https://our.internmc.facebook.com/intern/diff/D28318270/)

[ghstack-poisoned]

Commit:
Update on "[RPC Framework] Replace Python Pickler with internal RPC pickler for RemoteModule"


In order to support sending `RemoteModule` over PRC, previously the pickling/unpickling of `RemoteModule` was implemented based on `__setstate__` and `__getstate__`. However, this means that the user can call regular Python pickler/unpickler to invoke the same logic,which should not be allowed.

This PR ensures that the pickling can only happen over RPC and not via regular python pickle.

Additionally, when a new attribute is added to `RemoteModule`, if it's not added to either `_REMOTE_MODULE_PICKLED_ATTRIBUTES` or er`_REMOTE_MODULE_ATTRIBUTES_IGNORE_FOR_PICKLING`, this attribute will be ignored and an error message will be printed to std.err. However, it will not raise an exception like before, because such exception raised at the RPC layer won't be displayed properly.

#Closes: https://github.com/pytorch/pytorch/issues/57516

Differential Revision: [D28318270](https://our.internmc.facebook.com/intern/diff/D28318270/)

[ghstack-poisoned]

Commit:
Update on "[RPC Framework] Replace Python Pickler with internal RPC pickler for RemoteModule"


In order to support sending `RemoteModule` over PRC, previously the pickling/unpickling of `RemoteModule` was implemented based on `__setstate__` and `__getstate__`. However, this means that the user can call regular Python pickler/unpickler to invoke the same logic,which should not be allowed.

This PR ensures that the pickling can only happen over RPC and not via regular python pickle.

Additionally, when a new attribute is added to `RemoteModule`, if it's not added to either `_REMOTE_MODULE_PICKLED_ATTRIBUTES` or er`_REMOTE_MODULE_ATTRIBUTES_IGNORE_FOR_PICKLING`, this attribute will be ignored and an error message will be printed to std.err. However, it will not raise an exception like before, because such exception raised at the RPC layer won't be displayed properly.

#Closes: https://github.com/pytorch/pytorch/issues/57516

Differential Revision: [D28318270](https://our.internmc.facebook.com/intern/diff/D28318270/)

[ghstack-poisoned]

pritamdamania87 reviewed (commented):
Looks good! Just a few minor commments.

SciPioneer commented on torch/testing/_internal/distributed/nn/api/remote_module_test.py:
```diff
@@ -418,16 +419,44 @@ def test_send_remote_module_with_a_new_attribute_over_the_wire(self):
         ):
             new_attr_name = "new_attr"
             setattr(remote_module, new_attr_name, 1)
-            with self.assertRaisesRegex(
```
Can add it back once if timeout issue is not hit by raising the runtime error in `remote_module_recuder`.

SciPioneer commented on torch/distributed/rpc/internal.py:
```diff
@@ -83,6 +83,55 @@ def _script_module_reducer(self, script_module):
+        generated_methods,
+        module_rref_fork_data,
+    ):
+        m = object.__new__(dist.nn.RemoteModule)  # type: ignore[attr-defined]
```
`__new__` is rarely used. The only reference I can find in PyTorch codebase is https://github.com/pytorch/pytorch/blob/master/torch/distributed/distributed_c10d.py#L922

So I used the same style here.

Commit:
Update on "[RPC Framework] Replace Python Pickler with internal RPC pickler for RemoteModule"


In order to support sending `RemoteModule` over PRC, previously the pickling/unpickling of `RemoteModule` was implemented based on `__setstate__` and `__getstate__`. However, this means that the user can call regular Python pickler/unpickler to invoke the same logic,which should not be allowed.

This PR ensures that the pickling can only happen over RPC and not via regular python pickle.

Additionally, when a new attribute is added to `RemoteModule`, if it's not added to either `_REMOTE_MODULE_PICKLED_ATTRIBUTES` or er`_REMOTE_MODULE_ATTRIBUTES_IGNORE_FOR_PICKLING`, this attribute will be ignored and an error message will be printed to std.err. However, it will not raise an exception like before, because such exception raised at the RPC layer won't be displayed properly.

#Closes: https://github.com/pytorch/pytorch/issues/57516

Differential Revision: [D28318270](https://our.internmc.facebook.com/intern/diff/D28318270/)

[ghstack-poisoned]

Commit:
Update on "[RPC Framework] Replace Python Pickler with internal RPC pickler for RemoteModule"


In order to support sending `RemoteModule` over PRC, previously the pickling/unpickling of `RemoteModule` was implemented based on `__setstate__` and `__getstate__`. However, this means that the user can call regular Python pickler/unpickler to invoke the same logic,which should not be allowed.

This PR ensures that the pickling can only happen over RPC and not via regular python pickle.

Additionally, when a new attribute is added to `RemoteModule`, if it's not added to either `_REMOTE_MODULE_PICKLED_ATTRIBUTES` or er`_REMOTE_MODULE_ATTRIBUTES_IGNORE_FOR_PICKLING`, this attribute will be ignored and an error message will be printed to std.err. However, it will not raise an exception like before, because such exception raised at the RPC layer won't be displayed properly.

#Closes: https://github.com/pytorch/pytorch/issues/57516

Differential Revision: [D28318270](https://our.internmc.facebook.com/intern/diff/D28318270/)

[ghstack-poisoned]

rohan-varma commented on torch/distributed/rpc/internal.py:
```diff
@@ -83,6 +83,55 @@ def _script_module_reducer(self, script_module):
+                pickled_attrs[k] = v
+            # Check if unpickled attributes are all in _REMOTE_MODULE_ATTRIBUTES_IGNORE_FOR_PICKLING.
+            elif k not in dist.nn._REMOTE_MODULE_ATTRIBUTES_IGNORE_FOR_PICKLING:  # type: ignore[attr-defined]
+                print(
+                    "The new attribute ``{}`` of RemoteModule is ignored during RPC pickling. "
+                    "To pickle this attribute, it must be either in ``_REMOTE_MODULE_PICKLED_ATTRIBUTES`` or "
+                    "``_REMOTE_MODULE_ATTRIBUTES_IGNORE_FOR_PICKLING``.".format(k),
+                    file=sys.stderr,
+                )
```
Do we know what happens here such that we don't get the error immediately anymore? AFAIK this should be on the synchronous path of RPC so we should still get the error ideally.

SciPioneer replied:
> Do we know what happens here such that we don't get the error immediately anymore? AFAIK this should be on the synchronous path of RPC so we should still get the error ideally.

Good question. Here I don't understand where the code path goes once an exception is raised. I can confirm that the receiver is not reached, and `dispatch_table` is not explicitly queried in PyTorch codebase, so I guess it only runs at the third-party package.

mrshenli replied:
I am a bit lost, the reducer is called in during serialize, right? So if it's the user main thread that wants to send the RemoteModule to a remote worker, the error will be thrown in the user thread. Shouldn't the user be able to see that immediately? Why it caused hang? And why it reached the deserialization side?

mrshenli replied:
Or is it because you are trying send a RemoteModule as a return value of an RPC, and the serialize crash killed that RPC thread?

SciPioneer replied:
> Shouldn't the user be able to see that immediately?

Yes, the error can be captured immediately on the main thread (worker 0 in this case).

> Why it caused hang? And why it reached the deserialization side?

I don't quite understand here. I have printed out all the communications between worker0 and worker1, and I can see that the only RPC affected is  `remote_attributes()` -- due to the exception raised at the serialization stage on worker0, no deserialization occurs the worker1.

All the communications with the exception are shown as below:
**worker0 -> worker1:**
1) `create_module` (the first RPC to create a remote module as the setup, which is not really relevant here)
2) `remote_attributes` **(this RPC raises an exception and stops at serialization phase on worker0, so no deserialization on worker1)**
3) `broadcast_to_followers`
4) `None`

**worker1 -> worker0:** 
1) `gather_to_leader`
2) output of `remote_attributes()`, which is a dictionary here. **(Due to the exception, no output will be generated on worker1 and sent back to worker0.)**
3) `None`


Therefore, we can say that the RPC of `remote_attributes` **DOES NOT** reach the deserialization side, although there are still some other communication that can still result in deserialization on worker1 (`None` and possibly `broadcast_to_followers`).


The following is the error message after all the above communications are completed. It seems that worker1 is still listening to worker0 after all the above communications. Do you have any idea?
```
/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/common_distributed.py:541: Process 0 timed out with traceback: 

Current thread 0x00007f3578fd7700 (most recent call first):
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/common_distributed.py", line 476 in _event_listener
  File "/usr/local/fbcode/platform009/lib/python3.8/threading.py", line 870 in run
  File "/usr/local/fbcode/platform009/lib/python3.8/threading.py", line 932 in _bootstrap_inner
  File "/usr/local/fbcode/platform009/lib/python3.8/threading.py", line 890 in _bootstrap

Thread 0x00007f3690940740 (most recent call first):
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/distributed/rpc/api.py", line 312 in shutdown
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/distributed/rpc/api.py", line 79 in wrapper
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/dist_utils.py", line 103 in new_test_method
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/common_distributed.py", line 393 in wrapper
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/common_distributed.py", line 509 in run_test
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/common_distributed.py", line 496 in _run
  File "/usr/local/fbcode/platform009/lib/python3.8/multiprocessing/process.py", line 108 in run
  File "/usr/local/fbcode/platform009/lib/python3.8/multiprocessing/process.py", line 315 in _bootstrap
  File "/usr/local/fbcode/platform009/lib/python3.8/multiprocessing/spawn.py", line 129 in _main
  File "/usr/local/fbcode/platform009/lib/python3.8/multiprocessing/spawn.py", line 116 in spawn_main
  File "<string>", line 1 in <module>

/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/common_distributed.py:541: Process 1 timed out with traceback: 

Current thread 0x00007f0ba3fff700 (most recent call first):
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/common_distributed.py", line 476 in _event_listener
  File "/usr/local/fbcode/platform009/lib/python3.8/threading.py", line 870 in run
  File "/usr/local/fbcode/platform009/lib/python3.8/threading.py", line 932 in _bootstrap_inner
  File "/usr/local/fbcode/platform009/lib/python3.8/threading.py", line 890 in _bootstrap

Thread 0x00007f0cc7df2740 (most recent call first):
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/distributed/rpc/api.py", line 312 in shutdown
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/distributed/rpc/api.py", line 79 in wrapper
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/dist_utils.py", line 103 in new_test_method
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/common_distributed.py", line 393 in wrapper
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/common_distributed.py", line 509 in run_test
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/common_distributed.py", line 496 in _run
  File "/usr/local/fbcode/platform009/lib/python3.8/multiprocessing/process.py", line 108 in run
  File "/usr/local/fbcode/platform009/lib/python3.8/multiprocessing/process.py", line 315 in _bootstrap
  File "/usr/local/fbcode/platform009/lib/python3.8/multiprocessing/spawn.py", line 129 in _main
  File "/usr/local/fbcode/platform009/lib/python3.8/multiprocessing/spawn.py", line 116 in spawn_main
  File "<string>", line 1 in <module>

Timing out after 100 seconds and killing subprocesses.
ERROR
```

mrshenli commented on torch/distributed/rpc/internal.py:
```diff
@@ -259,3 +303,17 @@ def _start_record_function(exec_type, func_name, current_worker_name, dest_worke
 RemoteException = collections.namedtuple("RemoteException", ["msg", "exception_type"])
+# FIXME: Directly referencing `dist.nn._REMOTE_MODULE_PICKLED_ATTRIBUTES` here can cause
+# "AttributeError: module 'torch.distributed' has no attribute 'nn'".
+# RemoteModule = collections.namedtuple("RemoteModule", dist.nn._REMOTE_MODULE_PICKLED_ATTRIBUTES)
```
This looks like a hack. Besides, IIUC, `RemoteModule` is a layer on top of RPC/RRef. Referencing `RemoteModule` in RPC source code looks a bit weird. 

Is it possible to let RPC pickler expose an private API for other modules to register their reducers. And RemoteModule call that API to add itself?

SciPioneer replied:
> This looks like a hack. Besides, IIUC, `RemoteModule` is a layer on top of RPC/RRef. Referencing `RemoteModule` in RPC source code looks a bit weird.

This looks like a hack. 
Oh, this indeed a hack, by resolving the suggestion from @lw and avoid hard-coding the fields in reducer/receiver side.
 
> Is it possible to let RPC pickler expose a private API for other modules to register their reducers. And RemoteModule call that API to add itself?
Good idea. Let me try that this approach.

SciPioneer replied:
Thanks for the suggestion! Created a private API to register reducers. Now `RemoteModule` is not imported to `internal.py`.

Commit:
Update on "[RPC Framework] Replace Python Pickler with internal RPC pickler for RemoteModule"


In order to support sending `RemoteModule` over PRC, previously the pickling/unpickling of `RemoteModule` was implemented based on `__setstate__` and `__getstate__`. However, this means that the user can call regular Python pickler/unpickler to invoke the same logic,which should not be allowed.

This PR ensures that the pickling can only happen over RPC and not via regular python pickle.

Additionally, when a new attribute is added to `RemoteModule`, if it's not added to either `_REMOTE_MODULE_PICKLED_ATTRIBUTES` or er`_REMOTE_MODULE_ATTRIBUTES_IGNORE_FOR_PICKLING`, this attribute will be ignored and an error message will be printed to std.err. However, it will not raise an exception like before, because such exception raised at the RPC layer will somehow cause timeout.

#Closes: https://github.com/pytorch/pytorch/issues/57516

Differential Revision: [D28318270](https://our.internmc.facebook.com/intern/diff/D28318270/)

[ghstack-poisoned]

Commit:
Update on "[RPC Framework] Replace Python Pickler with internal RPC pickler for RemoteModule"


In order to support sending `RemoteModule` over PRC, previously the pickling/unpickling of `RemoteModule` was implemented based on `__setstate__` and `__getstate__`. However, this means that the user can call regular Python pickler/unpickler to invoke the same logic,which should not be allowed.

This PR ensures that the pickling can only happen over RPC and not via regular python pickle.

Additionally, when a new attribute is added to `RemoteModule`, if it's not added to either `_REMOTE_MODULE_PICKLED_ATTRIBUTES` or er`_REMOTE_MODULE_ATTRIBUTES_IGNORE_FOR_PICKLING`, this attribute will be ignored and an error message will be printed to std.err. However, it will not raise an exception like before, because such exception raised at the RPC layer will somehow cause timeout.

#Closes: https://github.com/pytorch/pytorch/issues/57516

Differential Revision: [D28318270](https://our.internmc.facebook.com/intern/diff/D28318270/)

[ghstack-poisoned]

Commit:
Update on "[RPC Framework] Replace Python Pickler with internal RPC pickler for RemoteModule"


In order to support sending `RemoteModule` over PRC, previously the pickling/unpickling of `RemoteModule` was implemented based on `__setstate__` and `__getstate__`. However, this means that the user can call regular Python pickler/unpickler to invoke the same logic,which should not be allowed.

This PR ensures that the pickling can only happen over RPC and not via regular python pickle.

Additionally, when a new attribute is added to `RemoteModule`, if it's not added to either `_REMOTE_MODULE_PICKLED_ATTRIBUTES` or `_REMOTE_MODULE_ATTRIBUTES_IGNORE_FOR_PICKLING`, this attribute will be ignored and an error message will be printed to std.err. However, it will not raise an exception like before, because such exception raised at the RPC layer will somehow cause timeout.

#Closes: https://github.com/pytorch/pytorch/issues/57516

Differential Revision: [D28318270](https://our.internmc.facebook.com/intern/diff/D28318270/)

[ghstack-poisoned]

mrshenli commented on torch/distributed/rpc/internal.py:
```diff
@@ -42,6 +41,13 @@ def __init__(self):
+        # Used for registering customized picklers.
+        self._class_reducer_dict = {}
+
+    def _register_reducer(self, obj_class, reducer):
```
This looks much better! Thanks!
